### PR TITLE
Add R.pathOr

### PR DESCRIPTION
**Pull request by jimf**
Implement R.pathOr as per the invitation for PR in #1054.

### CODE REVIEW

**Comment by jimf** [comment]
NOTE: I've more or less re-implemented R.path here, returning the default value in place of `undefined`. I'm thinking it would be preferable to instead leverage R.path directly and return the default if the return value was `undefined`, but I'm open to suggestions. New to this codebase.

**Comment by davidchambers** [comment]
Thanks for the patch, @jimf!

File: src/pathOr.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+var _curry3 = require('./internal/_curry3');
+
+
+/**
+ * If the given, non-null object has an own property at the given path, returns
+ * the value at that path. Otherwise returns the provided default value.
```
Comment (davidchambers) [inline src/pathOr.js]: We're doing `val[paths[idx]]`, so the property needn't be an *own* property. Let's update the description accordingly.

File: src/pathOr.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * @func
+ * @memberOf R
+ * @category Object
+ * @sig a -> [String] -> {k: v} -> v | a
```
Comment (davidchambers) [inline src/pathOr.js]: I suggest `a -> [String] -> Object -> a` as the type signature. The default value should be of the same type as the expected value.

File: src/pathOr.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * @param {*} d The default value.
+ * @param {Array} path The path to use.
+ * @return {*} The data at `path`.
+ * @example
+ *
+ *      R.pathOr('N/A', ['a', 'b'], {a: {b: 2}}); //=> 2
+ *      R.pathOr('N/A', ['a', 'b'], {c: {b: 2}}); //=> "N/A"
+ */
+module.exports = _curry3(function path(d, paths, obj) {
```
Comment (davidchambers) [inline src/pathOr.js]: It seems to me the second parameter should be named `path`. There is, after all, only one path.

Comment (davidchambers) [inline src/pathOr.js]: Also, the name of the function should be `pathOr`.

File: src/pathOr.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  if (obj == null) {
+    return d;
+  } else {
+    var val = obj;
+    for (var idx = 0, len = paths.length; idx < len && val != null; idx += 1) {
+      val = val[paths[idx]];
+    }
```
Comment (davidchambers) [inline src/pathOr.js]: Ramda style favours `while` loops over `for` loops. Also, we prefer not to cache array lengths unless necessary. We might write:

```javascript
var val = obj;
var idx = 0;
while (val != null && idx < path.length) {
  val = val[path[idx]];
  idx += 1;
}
```

File: src/pathOr.js (hunk 1 above)
Comment (jimf) [inline src/pathOr.js]: +1

File: src/pathOr.js (hunk 2 above)
Comment (jimf) [inline src/pathOr.js]: +1, that makes sense.

File: src/pathOr.js (hunk 3 above)
Comment (jimf) [inline src/pathOr.js]: +1 on `path` vs `paths`. I was just staying consistent with `R.path`'s implementation. As for the function name, doh!

File: src/pathOr.js (hunk 4 above)
Comment (jimf) [inline src/pathOr.js]: +1, will change.

**Comment by jimf** [comment]
@davidchambers your comments have been addressed.

A couple of your comments would apply to the `R.path` implementation as well. Thoughts on making those same mods there, either as part of this PR or another?

**Comment by raine** [comment]
I'm wondering about the reasons against implementing this in terms of `path` and `defaultTo` internally.

Are we trying to keep functions independent in case they are deprecated?

**Comment by davidchambers** [comment]
> I'm wondering about the reasons against implementing this in terms of `path` and `defaultTo` internally.

It seems we could use these functions to avoid duplicating code. Mind trying that, @jimf?

> A couple of your comments would apply to the `R.path` implementation as well. Thoughts on making those same mods there, either as part of this PR or another?

Let's include the `R.path` tweaks in this pull request.

**Comment by jimf** [comment]
+1, fully agree.

**Comment by jimf** [comment]
@davidchambers @raine comments addressed!

File: src/pathOr.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * If the given, non-null object has a value at the given path, returns
+ * the value at that path. Otherwise returns the provided default value.
+ *
+ * @func
+ * @memberOf R
+ * @category Object
+ * @sig a -> [String] -> Object -> a
+ * @param {*} d The default value.
+ * @param {Array} p The path to use.
+ * @return {*} The data at `path`.
```
Comment (davidchambers) [inline src/pathOr.js]: This description is not entirely accurate. Let's either reword or remove it.

**Comment by davidchambers** [comment]
There's just one minor issue for you to address, @jimf, before we merge this pull request.

**Comment by jimf** [comment]
@davidchambers Updated!

**Comment by jimf** [comment]
Regarding my second commit, I just want to resurface #1192 and make sure we still want that.

**Comment by davidchambers** [comment]
> Regarding my second commit, I just want to resurface #1192 and make sure we still want that.

We're in favour of `while` loops over `for` loops, and we like to avoid manual `.length` caching unless there's a possibility of the value changing during iteration. In this case a `while` loop without manual `.length` caching exactly matches our preferences. Thank you for fixing the style violations. :)

Please squash the commits (or append an "s" to the username mentioned in the second commit message).

**Comment by jimf** [comment]
@davidchambers Thanks for the clarification. Commits have been reworded and squashed.

**Comment by davidchambers** [comment]
Thanks! I'll merge as soon as Travis gives the go-ahead.

**Comment by davidchambers** [comment]
:deciduous_tree:

**Comment by jimf** [comment]
Thanks! :beers:

**Pull request merged**